data_prep/project_context/context_introspector.py

```diff
@@ -0,0 +1,203 @@
+
+    return reconstructed_type
+
+  def get_embeddable_types(self) -> list[str]:
+    """Retrieve types from FI."""
+    seen_types = set()
+    types_to_get = set()
+    types = set()
+
+    types_to_get.add(self._clean_type(self._benchmark.return_type))
+
+    params = self._benchmark.params
+
+    for param in params:
+      cleaned_type = self._clean_type(param['type'])
+      if not cleaned_type:
+        continue
+      types_to_get.add(cleaned_type)
+
+    seen_types = types_to_get.copy()
+
+    # Add support for recursively querying for types
+    iteration = 0
+    while types_to_get:
+      print("Iteration: {} Seen types: {}".format(iteration, seen_types))
+      iteration += 1
+      current_type = types_to_get.pop()
+      print(f'Querying for type: {current_type}')
+      info = introspector.query_introspector_type_info(self._benchmark.project,
+                                                       current_type)
+      if not info:
+        continue
+
+      reconstructed_type = ''
+      # For now, info is a single element and not a list.
+      # Requested it be changed to a list, so that type
+      # elements with the same name will be returned.
+      if info['type'] == 'struct':
+        reconstructed_type = self._get_struct_type(info, seen_types,
+                                                   types_to_get)
+      if info['type'] == 'typedef':
+        reconstructed_type = self._get_typedef_type(info)
+
+      if reconstructed_type:
+        types.add(reconstructed_type)
+
+    return list(types)
+
+  def _extract_type_from_source_line(self, source_line: str,
+                                     type_element: dict) -> str:
+    """Attempts to extract a type from a source line.
+    Do so by attempting to match for name and extracting
+    everything before it."""
+    start = source_line.find(next(filter(str.isalpha, source_line)))
+    end = source_line.find(type_element['name'])
+
+    if end == -1:
+      return ''
+
+    return source_line[start:end]
+
+  def _clean_type(self, type_name: str) -> str:
+    """Cleans a type so that it can be fetched from FI."""
+    if not type_name:
+      return type_name
+
+    if '*' in type_name:
+      type_name = type_name.replace('*', '')
+
+    type_tokens = type_name.split(' ')
+
+    # Could be a trailing space after the pointer is removed
+    if '' in type_tokens:
+      type_tokens.remove('')
+
+    if 'struct' in type_tokens:
+      type_tokens.remove('struct')
+
+    if 'enum' in type_tokens:
+      type_tokens.remove('enum')
+
+    if 'const' in type_tokens:
+      type_tokens.remove('const')
+
+    if 'volatile' in type_tokens:
+      type_tokens.remove('volatile')
+
+    # If there is more than a single token
+    # we probably do not care about querying for the type (?)
+    # E.g. unsigned [...], long [...], short [...], ...
+    # as they're most likely builtin.
+    if len(type_tokens) > 1:
+      return ''
+
+    return type_tokens[0]
```

Comment:
I presume this will be delegated to FI later?

Reply:
ideally this isn't required yup. currently parameter types in the benchmark yaml files look like `struct x` though and only `x` is expected during the FI lookup.